### ISSUE TITLE
Tracking issue for supported ONNX operators

### ISSUE THREAD

**Issue by robertknight** [robertknight/rten#14]
This is a tracking issue listing which [ONNX operators](https://onnx.ai/onnx/operators/) are currently supported.

Some things to note:

- Support for an operator does not mean that all attributes or data types listed in the current spec are supported.
- Some operators require additional dependencies to support. These typically require enabling additional crate features. For example the `Random*` ops require enabling the `random` crate feature
- Some operators are deprecated in the spec. Their non-deprecated replacements are implemented. This includes: Scatter, Upsample.
- Operators are usually implemented after finding a model that needs them. These models then serve as an initial test case. If you need an operator which is not currently listed as supported, it is helpful (but not essential) if you can point to an open source ONNX model which needs it.

<details>
  <summary>Script used to generate list</summary>

```py
from bs4 import BeautifulSoup
import requests

# URL of the ONNX operators page
url = "https://onnx.ai/onnx/operators/"

# Path to FlatBuffers schema listing supported ops.
schema_path = "src/schema.fbs"

# Fetch ONNX operators page and extract the list of operators.
#
# This assumes the operators are listed in the first table on the page and
# that the operator name is the first column.
response = requests.get(url)
soup = BeautifulSoup(response.content, 'html.parser')
operator_names = []

table = soup.find('table')
rows = table.find_all('tr')
for row in rows:
    cols = row.find_all('td')
    if len(cols) >= 1:
        operator_name = cols[0].text.strip()
        operator_names.append(operator_name)

# Scan FlatBuffers schema and extract supported operator names
supported_ops = set()
with open(schema_path) as fp:
    in_operator_type_enum = False
    for line in fp:
        if line.startswith('enum OperatorType'):
            in_operator_type_enum = True
            continue
        if line.startswith('}') and in_operator_type_enum:
            break

        if in_operator_type_enum:
            op_name = line.strip().replace(',', '')
            supported_ops.add(op_name)

# List operators and support status
for operator in operator_names:
    if operator in supported_ops:
        print(f"- [x] {operator}")
    else:
        print(f"- [ ] {operator}")
```
</details>

## Operator list

- [x] Abs
- [x] Acos
- [ ] Acosh
- [x] Add
- [ ] AffineGrid
- [x] And
- [x] ArgMax
- [x] ArgMin
- [x] Asin
- [ ] Asinh
- [x] Atan
- [ ] Atanh
- [x] AveragePool
- [x] BatchNormalization
- [ ] Bernoulli
- [ ] BitShift
- [ ] BitwiseAnd
- [ ] BitwiseNot
- [ ] BitwiseOr
- [ ] BitwiseXor
- [ ] BlackmanWindow
- [x] Cast
- [x] CastLike
- [x] Ceil
- [ ] Celu
- [ ] CenterCropPad
- [x] Clip
- [ ] Col2Im
- [ ] Compress
- [x] Concat
- [x] ConcatFromSequence (https://github.com/robertknight/rten/pull/859)
- [x] Constant (N/A - Constants are handled at model conversion time)
- [x] ConstantOfShape
- [x] Conv
- [x] ConvInteger
- [x] ConvTranspose
- [x] Cos
- [ ] Cosh
- [x] CumSum
- [ ] DFT
- [ ] DeformConv
- [x] DepthToSpace (https://github.com/robertknight/rten/pull/468)
- [x] DequantizeLinear
- [ ] Det
- [x] Div
- [x] Dropout (https://github.com/robertknight/rten/pull/652)
- [x] DynamicQuantizeLinear
- [x] Einsum
- [x] Elu
- [x] Equal
- [x] Erf
- [x] Exp
- [x] Expand
- [x] EyeLike (https://github.com/robertknight/rten/pull/741)
- [x] Flatten
- [x] Floor
- [x] GRU
- [x] Gather
- [x] GatherElements
- [x] GatherND
- [x] Gelu
- [x] Gemm
- [x] GlobalAveragePool
- [ ] GlobalLpPool
- [x] GlobalMaxPool (https://github.com/robertknight/rten/pull/1072)
- [x] Greater
- [x] GreaterOrEqual
- [x] GridSample (https://github.com/robertknight/rten/pull/867)
- [ ] GroupNormalization
- [ ] HammingWindow
- [ ] HannWindow
- [x] HardSigmoid
- [x] HardSwish
- [ ] Hardmax
- [x] Identity
- [x] If
- [ ] ImageDecoder
- [x] InstanceNormalization
- [x] IsInf (https://github.com/robertknight/rten/pull/837)
- [x] IsNaN (https://github.com/robertknight/rten/pull/837)
- [ ] LRN
- [x] LSTM
- [x] LayerNormalization
- [x] LeakyRelu
- [x] Less
- [x] LessOrEqual
- [x] Log
- [x] LogSoftmax
- [x] Loop (https://github.com/robertknight/rten/pull/855)
- [ ] LpNormalization
- [ ] LpPool
- [x] MatMul
- [x] MatMulInteger
- [x] Max
- [x] MaxPool
- [ ] MaxRoiPool
- [ ] MaxUnpool
- [x] Mean
- [ ] MeanVarianceNormalization
- [ ] MelWeightMatrix
- [x] Min
- [ ] Mish
- [x] Mod
- [x] Mul
- [ ] Multinomial
- [x] Neg
- [ ] NegativeLogLikelihoodLoss
- [x] NonMaxSuppression
- [x] NonZero
- [x] Not
- [x] OneHot
- [ ] Optional
- [ ] OptionalGetElement
- [ ] OptionalHasElement
- [x] Or
- [x] PRelu (https://github.com/robertknight/rten/pull/888)
- [x] Pad
- [x] Pow
- [ ] QLinearConv
- [ ] QLinearMatMul
- [x] QuantizeLinear
- [ ] RNN
- [x] RandomNormal
- [x] RandomNormalLike
- [x] RandomUniform
- [x] RandomUniformLike
- [x] Range
- [x] Reciprocal
- [ ] ReduceL1
- [x] ReduceL2
- [ ] ReduceLogSum
- [ ] ReduceLogSumExp
- [x] ReduceMax
- [x] ReduceMean
- [x] ReduceMin
- [x] ReduceProd
- [x] ReduceSum
- [x] ReduceSumSquare
- [ ] RegexFullMatch
- [x] Relu
- [x] Reshape
- [x] Resize
- [ ] ReverseSequence
- [ ] RoiAlign
- [x] Round
- [x] STFT (https://github.com/robertknight/rten/pull/921)
- [ ] Scan
- [ ] Scatter (_deprecated_. The replacement _ScatterElements_ is implemented)
- [x] ScatterElements
- [x] ScatterND
- [ ] Selu
- [x] SequenceAt (https://github.com/robertknight/rten/pull/858)
- [x] SequenceConstruct (https://github.com/robertknight/rten/pull/864)
- [x] SequenceEmpty (https://github.com/robertknight/rten/pull/858)
- [x] SequenceErase (https://github.com/robertknight/rten/pull/865)
- [x] SequenceInsert (https://github.com/robertknight/rten/pull/858)
- [x] SequenceLength (https://github.com/robertknight/rten/pull/863)
- [ ] SequenceMap
- [x] Shape
- [ ] Shrink
- [x] Sigmoid
- [x] Sign
- [x] Sin
- [ ] Sinh
- [x] Size
- [x] Slice
- [x] Softmax
- [ ] SoftmaxCrossEntropyLoss
- [x] Softplus
- [ ] Softsign
- [ ] SpaceToDepth
- [x] Split
- [x] SplitToSequence (https://github.com/robertknight/rten/pull/860)
- [x] Sqrt
- [x] Squeeze
- [ ] StringConcat
- [ ] StringNormalizer
- [ ] StringSplit
- [x] Sub
- [x] Sum
- [x] Tan
- [x] Tanh
- [ ] TfIdfVectorizer
- [ ] ThresholdedRelu
- [x] Tile
- [x] TopK
- [x] Transpose
- [x] Trilu
- [ ] Unique
- [x] Unsqueeze
- [ ] Upsample (deprecated. The replacement _Resize_ is implemented)
- [x] Where
- [x] Xor

**Comment by Caellian** [comment]
It would be cool if you could pin this issue, and link to it from README. For a newcomer, it was easier to gauge the state of the project (and that I didn't enable a feature) than going through schema currently linked to in the README.

**Comment by robertknight** [comment]
Hello @Caellian, which operators did you run into issues with. Given that you mentioned enabling a feature, I'm guessing it might be the `Random*` ops? There is [an issue](https://github.com/robertknight/rten/issues/154) about producing a more helpful error when an operator is supported, but not enabled due to the current crate features.

**Comment by Caellian** [comment]
> I'm guessing it was one of the Random* ops?

Yup, it was `RandomNormalLike`, figured it out by searching the code for it and seeing `cfg` attribute on one use, after seeing it marked as supported here.

**Comment by robertknight** [comment]
I asked Claude (Fable 5) to estimate the work involved in each of the remaining operators, as well as the cost in terms of code size and new dependencies. These are its estimates: https://gist.github.com/robertknight/08cd47a0251e0cd5953c71c575522850.
